internal/raft/cluster.go

```diff
@@ -0,0 +1,201 @@
+package raft
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jpillora/backoff"
+	"github.com/lni/dragonboat/v4"
+	"github.com/lni/dragonboat/v4/client"
+	"github.com/lni/dragonboat/v4/config"
+	"github.com/lni/dragonboat/v4/statemachine"
+)
+
+type RaftConfig struct {
+	InitialMembers []string `help:"Initial members" required:""`
+	ReplicaID      uint64   `help:"Node ID" required:""`
+	DataDir        string   `help:"Data directory" required:""`
+	RaftAddress    string   `help:"Address to advertise to other nodes" required:""`
+	ListenAddress  string   `help:"Address to listen for incoming traffic. If empty, RaftAddress will be used."`
+	// Raft configuration
+	ElectionRTT        uint64 `help:"Election RTT" default:"10"`
+	HeartbeatRTT       uint64 `help:"Heartbeat RTT" default:"1"`
+	SnapshotEntries    uint64 `help:"Snapshot entries" default:"10"`
+	CompactionOverhead uint64 `help:"Compaction overhead" default:"100"`
+}
+
+// Cluster of dragonboat nodes.
+type Cluster struct {
+	config *RaftConfig
+	nh     *dragonboat.NodeHost
+	shards map[uint64]statemachine.CreateStateMachineFunc
+}
+
+// ShardHandle is a handle to a shard in the cluster.
+// It is the interface to update and query the state of a shard.
+//
+// E is the event type.
+// Q is the query type.
+// R is the query response type.
+type ShardHandle[E Event, Q any, R any] struct {
+	shardID uint64
+	cluster *Cluster
+	session *client.Session
+}
+
+// Propose an event to the shard.
+func (s *ShardHandle[E, Q, R]) Propose(ctx context.Context, msg E) error {
+	if s.cluster.nh == nil {
```

Comment:
This probably needs a lock too?

Reply:
Does it? It should be only set once at `Start`